rust/src/http_client.rs

```diff
@@ -301,3 +307,23 @@ where
 
     Ok(deferred)
 }
+
+static MAKE_DEFERRED_YIELDABLE: OnceLock<pyo3::Py<pyo3::PyAny>> = OnceLock::new();
+
+/// Given a deferred, make it follow the Synapse logcontext rules
+fn make_deferred_yieldable<'py>(
+    py: Python<'py>,
+    deferred: &Bound<'py, PyAny>,
+) -> Bound<'py, PyAny> {
+    let make_deferred_yieldable = MAKE_DEFERRED_YIELDABLE.get_or_init(|| {
+        let sys = PyModule::import(py, "synapse.logging.context").unwrap();
+        let func = sys.getattr("make_deferred_yieldable").unwrap().unbind();
+        func
+    });
+
+    make_deferred_yieldable
+        .call1(py, (deferred,))
+        .unwrap()
+        .extract(py)
+        .unwrap()
+}
```

Comment:
This is based on @reivilibre suggestion in https://github.com/element-hq/synapse/pull/18903#discussion_r2332607994

But please double-check that I didn't fumble the lifetimes, etc trying to kludge things together.

Reply:
Ideally we'd avoid `unwrap` as much as possible, especially when a function call can potentially raise. Unfortunately [`OnceCell::get_or_try_init` is unstable](https://doc.rust-lang.org/stable/std/sync/struct.OnceLock.html#method.get_or_try_init) so we can't avoid it here easily, but the call to `make_deferred_yieldable` can bubble up errorsa

```suggestion
/// Given a deferred, make it follow the Synapse logcontext rules
fn make_deferred_yieldable<'py>(
    py: Python<'py>,
    deferred: &Bound<'py, PyAny>,
) -> PyErr<Bound<'py, PyAny>> {
    let make_deferred_yieldable = MAKE_DEFERRED_YIELDABLE.get_or_init(|| {
        let sys = PyModule::import(py, "synapse.logging.context").unwrap();
        let func = sys.getattr("make_deferred_yieldable").unwrap().unbind();
        func
    });

    make_deferred_yieldable
        .call1(py, (deferred,))?
        .extract(py)
}
```

Reply:
Had to adjust it a bit. Updated to this:

```rust
static MAKE_DEFERRED_YIELDABLE: OnceLock<pyo3::Py<pyo3::PyAny>> = OnceLock::new();

/// Given a deferred, make it follow the Synapse logcontext rules
fn make_deferred_yieldable<'py>(
    py: Python<'py>,
    deferred: &Bound<'py, PyAny>,
) -> PyResult<Bound<'py, PyAny>> {
    let make_deferred_yieldable = MAKE_DEFERRED_YIELDABLE.get_or_init(|| {
        let sys = PyModule::import(py, "synapse.logging.context").unwrap();
        let func = sys.getattr("make_deferred_yieldable").unwrap().unbind();
        func
    });

    make_deferred_yieldable.call1(py, (deferred,))?.extract(py)
}
```